### PR TITLE
Bug fix for lint rule

### PR DESCRIPTION
**Pull request by prateekbh**
The `no-html-link-for-pages.js` rules fails for directory pattern `/foo/[bar]/...`. 
This PR fixes it and adds test for it.

## Bug
- [x] Integration tests added

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| buildDuration | 17s | 17.3s | ⚠️ +259ms |
| buildDurationCached | 3.8s | 3.8s | -5ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.138 | 3.125 | -0.01 |
| / avg req/sec | 796.6 | 800.1 | +3.5 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.966 | 2.047 | ⚠️ +0.08 |
| /error-in-render avg req/sec | 1271.54 | 1221.32 | ⚠️ -50.22 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.3 kB | 23.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 491 B | 491 B | ✓ |
| Overall change | 491 B | 491 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| index.html gzip | 540 B | 540 B | ✓ |
| link.html gzip | 551 B | 551 B | ✓ |
| withRouter.html gzip | 534 B | 534 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| buildDuration | 13.9s | 13.7s | -135ms |
| buildDurationCached | 5.9s | 5.7s | -199ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.111 | 3.121 | ⚠️ +0.01 |
| / avg req/sec | 803.5 | 800.98 | ⚠️ -2.52 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.968 | 1.942 | -0.03 |
| /error-in-render avg req/sec | 1270.09 | 1287.15 | +17.06 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.7 kB | 10.7 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68.1 kB | 68.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.87 kB | 2.87 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 500 B | 500 B | ✓ |
| Overall change | 500 B | 500 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| index.html gzip | 585 B | 585 B | ✓ |
| link.html gzip | 598 B | 598 B | ✓ |
| withRouter.html gzip | 578 B | 578 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: 9d97df609f999284bcfc0c83b48bfef3038a83ca

File: packages/eslint-plugin-next/lib/utils/url.js [hunk 1]
```diff
@@ -42,7 +42,11 @@ function getUrlFromPagesDirectories(urlPrefix, directories) {
           (url) => `^${normalizeURL(url)}$`
         )
     )
-  ).map((urlReg) => new RegExp(urlReg))
+  ).map((urlReg) => {
+    // /foo/[bar] filees fails without this.
+    urlReg = urlReg.replace(/\[/g, "\\[");
+    new RegExp(urlReg)
```
Comment (ijjk) [inline packages/eslint-plugin-next/lib/utils/url.js:48]:
```suggestion
    return new RegExp(urlReg)
```
Should we escape all regex characters here like we do here? https://github.com/vercel/next.js/blob/acf65f92dc7ce459a25ac5a9bc668300e8cff28b/packages/next/shared/lib/router/utils/route-regex.ts#L9-L11

Comment (prateekbh) [reply packages/eslint-plugin-next/lib/utils/url.js:48]: we definitely should. I'll give it a try and update this

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 09b3f7bcb946c3a77c93bb1b5e3cff99ee1e5411

**test/unit/eslint-plugin-next/no-html-link-for-pages.test.ts** 
- no-html-link-for-pages > invalid dynamic route

<details>
<summary>Expand output</summary>

  ● no-html-link-for-pages › invalid dynamic route

    assert.notEqual(received, expected)

    Expected value to not be equal to:
      undefined
    Received:
      undefined

    Message:
      No lint errors found.

    Difference:

    Compared values have no visual difference.

      154 |       report.message,
      155 |       "Do not use the HTML <a> tag to navigate to /. Use Link from 'next/link' instead. See: https://nextjs.org/docs/messages/no-html-link-for-pages."
    > 156 |     )
          |      ^
      157 |   })
      158 |
      159 |   it('invalid dynamic route', function () {

      at Object.<anonymous> (unit/eslint-plugin-next/no-html-link-for-pages.test.ts:156:25)

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| buildDuration | 12.9s | 12.8s | -63ms |
| buildDurationCached | 3s | 2.9s | -56ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | -4 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.331 | 2.3 | -0.03 |
| / avg req/sec | 1072.73 | 1086.73 | +14 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.24 | 1.24 | ✓ |
| /error-in-render avg req/sec | 2016.78 | 2015.51 | ⚠️ -1.27 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.3 kB | 23.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 491 B | 491 B | ✓ |
| Overall change | 491 B | 491 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| index.html gzip | 540 B | 540 B | ✓ |
| link.html gzip | 552 B | 552 B | ✓ |
| withRouter.html gzip | 533 B | 533 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| buildDuration | 10.5s | 10.5s | ⚠️ +42ms |
| buildDurationCached | 4.5s | 4.5s | -35ms |
| nodeModulesSize | 61.7 MB | 61.7 MB | -4 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.307 | 2.286 | -0.02 |
| / avg req/sec | 1083.55 | 1093.69 | +10.14 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.256 | 1.256 | ✓ |
| /error-in-render avg req/sec | 1991.07 | 1990.7 | ⚠️ -0.37 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.7 kB | 10.7 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68.1 kB | 68.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.87 kB | 2.87 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 500 B | 500 B | ✓ |
| Overall change | 500 B | 500 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | prateekbh/next.js lint-fix | Change |
| - | - | - | - |
| index.html gzip | 586 B | 586 B | ✓ |
| link.html gzip | 597 B | 597 B | ✓ |
| withRouter.html gzip | 578 B | 578 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: 09b3f7bcb946c3a77c93bb1b5e3cff99ee1e5411

File: packages/eslint-plugin-next/lib/utils/url.js (hunk 1 above)
Comment (prateekbh) [reply packages/eslint-plugin-next/lib/utils/url.js]: I moved the logic around to handle this, we don't need to handle the escaping of all the items

Comment (prateekbh) [reply packages/eslint-plugin-next/lib/utils/url.js]: added more tests to check all permutations and combinations

File: packages/eslint-plugin-next/lib/utils/url.js [hunk 2]
```diff
@@ -73,6 +75,9 @@ function parseUrlForPages(urlprefix, directory) {
     }
   })
   return res
+  // return res.map((expression) =>
+  //   expression.replace(/[|\\{}()[\]^$+*?.-]/g, '\\$&')
+  // )
```
Comment (housseindjirdeh) [inline packages/eslint-plugin-next/lib/utils/url.js:80]: Should this commented code be here?

**Review by housseindjirdeh — DISMISSED** [review]
Thanks for putting up a fix for this @prateekbh :)

LGTM (aside from the commented out code) ✅

File: packages/eslint-plugin-next/lib/utils/url.js (hunk 2 above)
Comment (prateekbh) [reply packages/eslint-plugin-next/lib/utils/url.js:80]: nope just waiting for the tests to pass

**Pull request merged**